File: src/main/java/net/joostory/jpastudy/ch08/Main.java
package net.joostory.jpastudy.ch08;

import javax.persistence.EntityManager;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;
import net.joostory.jpastudy.Entity_managerKt;

public class Main {
  private EntityManager em;

  public Main(EntityManager em) {
    this.em = em;
  }

  private void saveData() {
    Member2 member = new Member2();
    member.setUsername("회원1");
    em.persist(member);

    Team2 team = new Team2();
    team.setName("팀1");
    em.persist(team);
  }

  public void run() {
    saveData();
    em.flush();
    em.clear();
    referenceTeam();
  }

  private void referenceTeam() {
    Member2 member = em.find(Member2.class, 1L);
    Team2 team = em.getReference(Team2.class, 2L);
    System.out.println("team=" + team.getId());
    System.out.println("team=" + em.getEntityManagerFactory().getPersistenceUnitUtil().isLoaded(team));
    System.out.println("team=" + team.getClass().getName());
    member.setTeam(team);
  }

  private void criteriaTest() {
    CriteriaBuilder cb = em.getCriteriaBuilder();
    CriteriaQuery<Member2> query = cb.createQuery(Member2.class);
    Root<Member2> m = query.from(Member2.class);
    CriteriaQuery<Member2> cq = query.select(m).where(cb.equal(m.get("username"), "kim"));
    em.createQuery(cq).getResultList();
  }


  public static void main(String argv[]) {
    System.out.println("Chapter08 (Java)");
    Entity_managerKt.runWithEntityManager(em -> {
      Main m = new Main(em);
      m.run();
      return null;
    });
  }
}

File: src/main/kotlin/net/joostory/jpastudy/ch08/main.kt
package net.joostory.jpastudy.ch08

import net.joostory.jpastudy.log
import net.joostory.jpastudy.runWithEntityManager
import javax.persistence.EntityManager

private fun saveData(em: EntityManager) {
  val order = Order(
    orderAmount = 1000
  )

  val member = Member(
    username = "회원1"
  )
  order.member = member
  member.orders.add(order)
  em.persist(member)
  em.persist(Member(
    username = "회원2"
  ))
  val team = Team(
    name = "팀1"
  )
  em.persist(team)
  member.team = team
  em.persist(Team(
    name = "팀2"
  ))
}

private fun printUserAndTeam(em: EntityManager) {
  val member = em.find(Member::class.java, 1L)
  val team = member.team
  log("회원이름: ${member.username}")
  log("팀이름: ${team?.name}")
}

private fun printUser(em: EntityManager) {
  val member = em.find(Member::class.java, 1L)
  log("회원이름: ${member.username}")
}

private fun referenceMember(em: EntityManager) {
  val member = em.getReference(Member::class.java, 3L)
  val team = em.getReference(Team::class.java, 5L)
  log("isLoaded = ${em.entityManagerFactory.persistenceUnitUtil.isLoaded(team)}")
  log("proxy = ${team.javaClass.name}")
  log("연관관계 설정 ${team.id}")
  member.team = team
  log("연관관계 완료")
}

private fun printOrders(em: EntityManager) {
  val member = em.getReference(Member::class.java, 3L)
  log("orders = ${member.orders.javaClass.name}")
}

private fun deleteOrder(em: EntityManager) {
  val order = em.getReference(Order::class.java, 2L)
  em.remove(order)
}

fun main() {
  println("Chapter 8")
  runWithEntityManager {
    saveData(it)
    it.flush()
    it.clear()
    printUser(it)
    printUserAndTeam(it)
    referenceMember(it)
    printOrders(it)
    deleteOrder(it)
  }
}
